### PR TITLE
Introduce device capability flag and default handler for parameter broadcasting

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
This PR is the second one to introduce parameter broadcasting, following up on #2575.

**Description of the Change:**
Here we introduce the device capability flag `"supports_broadcasting"` and a tape expansion function `unbroadcast_expand` (a `batch_transform`) that expands broadcasted tapes into unbroadcasted ones. This function is then used in `device.batch_transform` to expand broadcasted tapes whenever the device does not support broadcasting.

**Benefits:**
- Devices know whether they support broadcasting.
- Devices can handle broadcasted tapes even if they do not support broadcasting, simply by expanding such tapes into multiple tapes.

**Possible Drawbacks:**
- (unlikely:) Performance decrease when using broadcasted tapes due to additional transform efforts.

### CODE REVIEW

**Review by rsalvo77 — CHANGES_REQUESTED** [review]
Please review

File: pennylane/_device.py [hunk 1]
```diff
@@ -705,11 +705,6 @@ def batch_transform(self, circuit):
             the sequence of circuits to be executed, and a post-processing function
             to be applied to the list of evaluated circuit results.
         """
-
-        # If the observable contains a Hamiltonian and the device does not
-        # support Hamiltonians, or if the simulation uses finite shots, or
-        # if the Hamiltonian explicitly specifies an observable grouping,
-        # split tape into multiple tapes of diagonalizable known observables.
```
Comment (josh146) [inline pennylane/_device.py:712]: This might cause some merge conflicts with #2587 (@Qottmann), something to be aware of for whoever merges second 😆

File: pennylane/_device.py [hunk 2]
```diff
@@ -721,16 +716,48 @@ def batch_transform(self, circuit):
...
+        batch_size = circuit.batch_size
+        supports_broadcasting = self.capabilities().get("supports_broadcasting")
+
+        if batch_size is None or supports_broadcasting:
+            # If the circuit wasn't broadcasted or broadcasting is supported, no action required
+            return circuits, hamiltonian_fn
+
+        unbroadcasted_circuits = []
+        unbroadcast_fns = []
+        # Expand each of the broadcasted Hamiltonian-expanded circuits
```
Comment (josh146) [inline pennylane/_device.py:745]:
```suggestion

        # Expand each of the broadcasted Hamiltonian-expanded circuits
```

File: pennylane/_device.py [hunk 3]
```diff
@@ -721,16 +716,48 @@ def batch_transform(self, circuit):
...
+        for c in circuits:
+            tapes_batch, fn = qml.transforms.unbroadcast_expand(c)
+            unbroadcasted_circuits.extend(tapes_batch)
+            unbroadcast_fns.append(fn)
+
+        # Chain the postprocessing functions of the broadcasted-tape expansions and the Hamiltonian
+        # expansion. Note that the application order is reversed compared to the expansion order
```
Comment (josh146) [inline pennylane/_device.py:752]: wait, how come?

File: pennylane/_device.py [hunk 4]
```diff
@@ -721,16 +716,48 @@ def batch_transform(self, circuit):
...
+            # If the circuit wasn't broadcasted or broadcasting is supported, no action required
+            return circuits, hamiltonian_fn
+
+        unbroadcasted_circuits = []
+        unbroadcast_fns = []
+        # Expand each of the broadcasted Hamiltonian-expanded circuits
+        for c in circuits:
+            tapes_batch, fn = qml.transforms.unbroadcast_expand(c)
+            unbroadcasted_circuits.extend(tapes_batch)
+            unbroadcast_fns.append(fn)
```
Comment (josh146) [inline pennylane/_device.py:749]: @dwierichs would

https://github.com/PennyLaneAI/pennylane/blob/bd552accef371b21bde3ed862a527ee140e31dfd/pennylane/transforms/batch_transform.py#L422

help avoid the need to manually loop over the circuit to apply the batch transform? It should allow you to just do

```python
return map_batch_transform(unbroadcast_expand, circuits)
```

File: pennylane/transforms/unbroadcast_expand.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the tape expansion function for unbroadcasting a
+broadcasted tape into multiple tapes."""
+import pennylane as qml
+from pennylane.transforms.batch_transform import batch_transform
```
Comment (josh146) [inline pennylane/transforms/unbroadcast_expand.py:17]:
```suggestion
from .batch_transform import batch_transform
```

Not entirely sure I remember, but I _think_ you want a relative import here?

File: pennylane/transforms/unbroadcast_expand.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+
+@batch_transform
+def unbroadcast_expand(tape):
```
Comment (josh146) [inline pennylane/transforms/unbroadcast_expand.py:21]: minor comment, but I must admit I find the name hard to parse! Would the following do a better job of describing what the transform does?

- `broadcast`
- `broadcast_parameters`
- `parameter_expand`

In my mind, broadcasting is almost the _act_ of taking one tape with ndim parameters and creating multiple tapes 🤔 So 'unbroadcasting' is hard for me to parse

File: pennylane/transforms/unbroadcast_expand.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    r"""Expand a broadcasted tape into multiple unbroadcasted tapes
+    and a function that stacks and squeezes the results.
+
+    Args:
+        tape (.QuantumTape): Broadcasted tape to be expanded
+    Returns:
```
Comment (josh146) [inline pennylane/transforms/unbroadcast_expand.py:27]:
```suggestion

    Returns:
```

File: pennylane/transforms/unbroadcast_expand.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+"""This module contains the tape expansion function for unbroadcasting a
+broadcasted tape into multiple tapes."""
+import pennylane as qml
+from pennylane.transforms.batch_transform import batch_transform
+
+
+@batch_transform
+def unbroadcast_expand(tape):
+    r"""Expand a broadcasted tape into multiple unbroadcasted tapes
+    and a function that stacks and squeezes the results.
```
Comment (josh146) [inline pennylane/transforms/unbroadcast_expand.py:23]: Yeah I think this is where I get confused. Broadcasting to me is a verb; a tape can have ndim parameters, but it can't be described as 'broadcasted' before anything happens. Broadcasting only happens when you expand 🤔

Comment (josh146) [reply pennylane/transforms/unbroadcast_expand.py:23]: In a way, the broadcasting _is_ the expansion?

File: pennylane/transforms/__init__.py [hunk 9]
```diff
@@ -224,3 +224,4 @@
 from .transpile import transpile
 from . import qcut
 from .qcut import cut_circuit, cut_circuit_mc
+from .unbroadcast_expand import unbroadcast_expand
```
Comment (josh146) [inline pennylane/transforms/__init__.py:227]: Don't forget to add it to the autosummary above so that it shows up in the docs :)

File: pennylane/transforms/unbroadcast_expand.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    Args:
+        tape (.QuantumTape): Broadcasted tape to be expanded
+    Returns:
+        list[.QuantumTape]: Unbroadcasted tapes that produce one of the
+        results of the broadcasted tape each
+        callable: Function that stacks the results
```
Comment (josh146) [inline pennylane/transforms/unbroadcast_expand.py:30]: Might need to double check this renders correctly once this shows up in the docs!

File: pennylane/transforms/unbroadcast_expand.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    >>> with qml.tape.QuantumTape() as tape:
+    >>>     qml.RX(np.array([0.2, 0.6, 1.0], requires_grad=True), wires=0)
+    >>>     qml.expval(qml.PauliZ(0))
+    >>> tapes, fn = qml.transforms.unbroadcast_expand(tape)
+    >>> tapes
+    [<QuantumTape: wires=[0], params=1>,
+     <QuantumTape: wires=[0], params=1>,
+     <QuantumTape: wires=[0], params=1>]
+    >>> fn(qml.execute(tapes, qml.device("default.qubit", wires=1), None))
+    array([0.98006658, 0.82533561, 0.54030231])
```
Comment (josh146) [inline pennylane/transforms/unbroadcast_expand.py:52]: Very nice+clean!

File: pennylane/transforms/unbroadcast_expand.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        raise ValueError("The provided tape is not broadcasted.")
+
+    # Note that these unbatched_params will have shape (#params, num_tapes)
+    unbatched_params = []
+    for op in tape.operations + tape.observables:
+        for j, p in enumerate(op.data):
+            if op.batch_size and qml.math.ndim(p) != op.ndim_params[j]:
+                unbatched_params.append(qml.math.unstack(p))
+            else:
+                unbatched_params.append([p] * num_tapes)
```
Comment (josh146) [inline pennylane/transforms/unbroadcast_expand.py:66]: oh, this is actually very neat!

File: tests/devices/test_default_qubit.py [hunk 13]
```diff
@@ -2527,3 +2528,76 @@ def test_error_hamiltonian_expval_finite_shots(self):
...
+        dev = qml.device("default.qubit", wires=2, shots=shots)
+
+        H = qml.Hamiltonian([0.3, 0.9], [qml.PauliZ(0), qml.PauliY(1)])
+        H.compute_grouping()
+
+        @qml.qnode(dev)
+        def circuit(x, y):
+            self.RX_batched(x, wires=0)
+            self.RX_batched(y, wires=1)
+            return qml.expval(H)
```
Comment (josh146) [inline tests/devices/test_default_qubit.py:2596]: is it possible to batch `H`?

Comment (josh146) [reply tests/devices/test_default_qubit.py:2596]: E.g., 

```python
H = qml.Hamiltonian([np.array([0.1, 0.2]), np.array([1., -4.]), [qml.PauliZ(0), qml.PauliY(1)])
```

File: tests/devices/test_default_qubit.py [hunk 14]
```diff
@@ -2527,3 +2528,76 @@ def test_error_hamiltonian_expval_finite_shots(self):
...
+
+        out = circuit(x, y)
+        expected = 0.3 * qml.math.cos(x) * qml.math.ones_like(y) - 0.9 * qml.math.sin(y)
+
+        assert circuit.device.num_executions == len(y)
+        tol = 1e-10 if shots is None else 1e-2
+        assert qml.math.allclose(out, expected, atol=tol, rtol=0)
```
Comment (josh146) [inline tests/devices/test_default_qubit.py:2603]: good idea to test this!

File: tests/transforms/test_unbroadcast_expand.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+import numpy as np
+import pennylane as qml
+from pennylane import numpy as pnp
+
+
+class RX_batched(qml.RX):
+    """A version of qml.RX that detects batching."""
+
+    ndim_params = (0,)
+    compute_decomposition = staticmethod(lambda theta, wires=None: qml.RX(theta, wires=wires))
```
Comment (josh146) [inline tests/transforms/test_unbroadcast_expand.py:25]: I suppose this implies we still need to go back and update all the ops? 😬

**Review by josh146 — CHANGES_REQUESTED** [review]
Very nice and clean PR @dwierichs! Mostly left questions, but don't forget to also (a) add this PR to the changelog, and (b) ensure the new docstring appears in the docs

File: pennylane/transforms/unbroadcast_expand.py (hunk 6 above)
Comment (dwierichs) [reply pennylane/transforms/unbroadcast_expand.py:21]: > In my mind, broadcasting is almost the act of taking one tape with ndim parameters and creating multiple tapes 
Ah I see! I'm not sure, which name fits best here. In docs and comments I have been using "broadcasted tape" for tapes with `batch_size!=None`. Maybe `broadcast_expand` actually is a good name? It makes clear that it is in the context of broadcasting and expands into multiple tapes. What do you think?

File: pennylane/transforms/unbroadcast_expand.py (hunk 8 above)
Comment (dwierichs) [reply pennylane/transforms/unbroadcast_expand.py:23]: This is a tough one regarding name space :thinking: 
>but it can't be described as 'broadcasted' before anything happens
I think my understanding differs at this point, leading to the confusion. `qml.RX([1, 2], wires=0)` in my understanding broadcasts the operator across the additional parameter axis, and I thought broadcasting the operator makes the operator broadcasted. Looking at NumPy for a comparison again, broadcasting would mean to me to do 
`np.ones((3, 4)) * np.ones(4)` whereas the result of this expansion would correspond to duplicating the operation, yielding `np.array([row * np.ones(4) for row in np.ones((3, 4))])`.

I'm not sure how to be clearer here? :thinking: Happy for any suggestions you might have :)

File: pennylane/transforms/unbroadcast_expand.py (hunk 12 above)
Comment (dwierichs) [reply pennylane/transforms/unbroadcast_expand.py:67]: I was surprised myself how nicely this works out :D

File: tests/devices/test_default_qubit.py (hunk 13 above)
Comment (dwierichs) [reply tests/devices/test_default_qubit.py:2596]: Not yet... I think! :D This will depend also on the next PRs that allow to actually access numerical representations of operators and execute circuits on devices...
I'd hope to make this possible but have to check with all the processing of Hamiltonians...

File: tests/transforms/test_unbroadcast_expand.py (hunk 15 above)
Comment (dwierichs) [reply tests/transforms/test_unbroadcast_expand.py:25]: Ah, yes, this will happen in the updated version of #2535 :)
Note that only the `ndim_params` will be necessary. The `compute_decomposition` addition is just required to make `RX_batched` executable on a device (via decomposition).

File: pennylane/_device.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/_device.py]: It was super useful! Reducing the addition to `Device.batch_transform` to a few lines :)

File: pennylane/transforms/unbroadcast_expand.py (hunk 5 above)
Comment (dwierichs) [reply pennylane/transforms/unbroadcast_expand.py]: This was a legacy decision from when I had put the transform in a different file and there were cyclic imports :D Thanks for the catch!

File: pennylane/transforms/unbroadcast_expand.py (hunk 10 above)
Comment (dwierichs) [reply pennylane/transforms/unbroadcast_expand.py]: It didn't. Changed the formatting to a tuple.

**Comment by dwierichs** [comment]
[sc-19789]

File: tests/devices/test_default_qubit.py (hunk 13 above)
Comment (josh146) [reply tests/devices/test_default_qubit.py:2596]: If I recall, this was a recent request of @cvjjm, hence why it was on my mind 😆

File: pennylane/_device.py [hunk 16]
```diff
@@ -733,31 +733,22 @@ def batch_transform(self, circuit):
...
-            ]
-            return hamiltonian_fn(intermediate)
-
-        return unbroadcasted_circuits, chained_processing
+        # expansion. Note that the application order is reversed compared to the expansion order,
+        # i.e. while we first applied `hamiltonian_expand` to the tape, we need to process the
+        # results from the broadcast expansion first.
+        total_processing = lambda results: hamiltonian_fn(expanded_fn(results))
+
+        return expanded_tapes, total_processing
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/_device.py:751]: Right? :D Loving it!

File: pennylane/transforms/unbroadcast_expand.py (hunk 8 above)
Comment (dwierichs) [reply pennylane/transforms/unbroadcast_expand.py]: I think then it should be fine, together with removing the inconvenient "unbroadcast" naming, right? :)

Comment (josh146) [reply pennylane/transforms/unbroadcast_expand.py]: yep!

File: pennylane/transforms/broadcast_expand.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    >>> with qml.tape.QuantumTape() as tape:
+    >>>     qml.RX(np.array([0.2, 0.6, 1.0], requires_grad=True), wires=0)
+    >>>     qml.expval(qml.PauliZ(0))
+    >>> tapes, fn = qml.transforms.broadcast_expand(tape)
+    >>> tapes
+    [<QuantumTape: wires=[0], params=1>,
+     <QuantumTape: wires=[0], params=1>,
+     <QuantumTape: wires=[0], params=1>]
+    >>> fn(qml.execute(tapes, qml.device("default.qubit", wires=1), None))
+    array([0.98006658, 0.82533561, 0.54030231])
```
Comment (josh146) [inline pennylane/transforms/broadcast_expand.py:53]: Only one minor suggestion @dwierichs --- could you show a QNode example first, before the tape example?

With the QNode example, you can also use `qml.draw()` to show the QNode producing multiple circuits :)

Comment (josh146) [reply pennylane/transforms/broadcast_expand.py:53]: The reason for this is just to 'overcommunicate' that this works with QNodes, otherwise we have developers dropping down to tapes when they don't need to

File: requirements-ci.txt [hunk 18]
```diff
@@ -12,4 +12,5 @@ semantic_version==2.6
 dask[delayed]
 autoray
 matplotlib
-opt_einsum
\ No newline at end of file
+opt_einsum
+protobuf<4.21.0
```
Comment (josh146) [inline requirements-ci.txt:16]: I think this can be removed now!

File: requirements.txt [hunk 19]
```diff
@@ -15,3 +15,4 @@ autoray==0.3.1
 matplotlib~=3.5
 black>=21
 opt_einsum~=3.3
+protobuf<4.21.0
```
Comment (josh146) [inline requirements.txt:18]: same here

File: requirements-ci.txt (hunk 18 above)
Comment (josh146) [reply requirements-ci.txt:16]: when you merge in master, just pull in whatever master has here :)

**Review by josh146 — APPROVED** [review]
Thanks @dwierichs!
